File: houseRobberII/houseRobberII.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        int[] nums1 = new int[nums.length - 1];
        int[] nums2 = new int[nums.length - 1];
        for (int i = 0; i < nums.length; i++) {
            if (i > 0) nums2[i - 1] = nums[i];
            if (i < nums.length - 1) nums1[i] = nums[i];
        }
        return Math.max(robHelper(nums1), robHelper(nums2));
    }
    
    public int robHelper(int[] nums) {
        if (nums.length < 2) return nums[0];
        int[] dp = new int[nums.length];
        dp[0] = nums[0];
        dp[1] = nums[1];
        if (nums.length == 2) return Math.max(dp[0], dp[1]);
        dp[2] = nums[0] + nums[2];
        int res = Math.max(dp[1], dp[2]);
        for (int i = 3; i < dp.length; i++) {
            dp[i] = Math.max(dp[i - 2], dp[i - 3]) + nums[i];
            res = Math.max(res, dp[i]);
        }
        return res;      
    }
}

File: 213.houseRobberII/houseRobberII.cpp
// clang++ houseRobberII.cpp -std=c++11

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
  int rob(vector<int>& nums) {
    int size = nums.size();
    if (size == 0)
      return 0;
    if (size == 1)
      return nums[0];
    return max(robHelper(nums, 0, size - 2), robHelper(nums, 1, size - 1));
  }
  int robHelper(vector<int>& nums, int start, int end) {
    if (end == start)
      return nums[start];
    vector<int> dp(end - start + 1, 0);
    dp[0] = nums[start];
    dp[1] = max(nums[start], nums[start + 1]);
    for (int i = 2; i <= end - start; i++) {
      dp[i] = max(nums[start + i] + dp[i - 2], dp[i - 1]);
    }
    return dp[end - start];
  }
};

int main()
{
  Solution s;
  vector<int> nums = {7,1,3,5,8,10};
  cout << s.rob(nums) << endl;
  return 0;
}
